fix: Better support container restarts

georglauterbach opened the pull request:
# Description

<!--
  Include a summary of the change.
  Please also include relevant motivation and context.
-->

<!-- Link the issue which will be fixed (if any) here: -->
Fixes #4108
Fixes #4150

ref #4302

## Type of change

<!-- Delete options that are not relevant. -->

- [x] Improvement (non-breaking change that does improve existing functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary, I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] **I have added information about changes made in this PR to `CHANGELOG.md`**

casperklein commented on target/scripts/start-mailserver.sh:
```diff
@@ -174,27 +182,18 @@ function _register_functions() {
-  _log 'info' "Welcome to docker-mailserver ${DMS_RELEASE}"
+_register_functions
+_check
+_setup
```
`_setup` is called, even when the container was restarted?

I would revert the 2nd commit. Using one `if` clause at the bottom instead of three, makes it easier to see, what's exactly done on a fresh container / restarted container.

polarathene replied:
Agreed, looks like we can just have a single conditional that calls `_setup` only, we just need to look at two of the setup registered functions being moved to the `_check` function group instead as they're more generic to sanitization/prep before starting services.

polarathene commented on target/scripts/start-mailserver.sh:
```diff
@@ -174,27 +182,18 @@ function _register_functions() {
+_early_variables_setup
 
-  _log 'info' "Welcome to docker-mailserver ${DMS_RELEASE}"
+[[ -f /CONTAINER_START ]] && _log 'info' 'Container was restarted. Skipping most setup routines.'
```
I don't find this as intuitive. You've got this conditional scattered in several parts of the file now whereas before it was a much easier to grok overview of this or that.

https://github.com/docker-mailserver/docker-mailserver/blob/3ebca5dabaea1469f482d22912e08ec33c16cfed/target/scripts/start-mailserver.sh#L177-L197

Let's go over that.

```
_early_supervisor_setup
_early_variables_setup
```

`_early` methods are used to run before anything else.
- I am not sure if the supervisor one really needs to be run early per-se, it just needs to be run before  `daemon-stack.sh` methods, which AFAIK are all handled via `_register_start_daemon()` calls during the later `_register_functions()`, followed by `_start_daemons()` that actually runs it all, given what `_early_supervisor_setup()` in `setup-stack.sh` actually does, you're right here to not conditionally run it but it could probably be relocated to `daemon-stack.sh` and called early in `_start_daemons()`?

The variables one is a bit more involved. This method already lives in `variables-stack.sh` ensuring that vars our scripts expect are all setup correctly ahead of time. Then the remaining methods of `variables-stack.sh` are mostly all called during `_register_functions()` rather than registering them. That seems to be for convenience to co-locate them with their respective feature flags (ldap, oauth, saslauthd).

The final `variables-stack.sh` method export vars and instead of being run directly during `_register_functions()` is registered as a `setup-stack.sh` managed function (_thus this vars method won't run during container restart which probably was a **bug**?_). Technically since restarts avoid calling `_setup()` and export outputs container state which I think setup methods adjust defaults for, along with some `setup` CLI methods sourcing to run, there's probably a conflict there where exporting could introduce breakage (_change to ENV or anything that would affect our `VARS` like `_check_log_level()`, should be less prone to introducing breakage, but generally should rely on a container recreation rather than restart_).

---

As per the associated issues you referenced, `_check()` / `check-stack.sh` is suitable as a prep step relevant to restarts. The existing check methods maybe less so, but `_setup_save_states()` and `_setup_apply_fixes_after_configuration()` are relevant preparation / sanitization steps.

Although I'm note entirely sure at a glance for `_setup_save_states()`, while correcting permissions is important, it may make some assumptions that may not be desired during a restart?:

https://github.com/docker-mailserver/docker-mailserver/blob/f0daa1c8abff606bd3fa6f5212b7e77c1e7b38a2/target/scripts/startup/setup.d/mail_state.sh#L43-L46

---

[`_run_user_patches()`](https://github.com/docker-mailserver/docker-mailserver/blob/f0daa1c8abff606bd3fa6f5212b7e77c1e7b38a2/target/scripts/startup/setup-stack.sh#L106) could probably be registered as a setup at the end of `_register_functions`?

We'd then have:

```bash
 _early_supervisor_setup
 _early_variables_setup

 _log 'info' "Welcome to docker-mailserver ${DMS_RELEASE}"

 _register_functions
 _check

 # Ensure DMS only adjusts config files for a new container.
 # Container restarts should skip as they retain the modified config.
 if [[ -f /CONTAINER_START ]]; then
   _log 'info' 'Container was restarted. Skipping setup routines.'
 else
   _setup
 fi
```

---

That seems better for maintenance?

- [ ] Revert the conditional approach you added to `_register_functions()` in favor of being DRY.
- [ ] Move `_run_user_patches` to `_register_functions()` as `_register_setup_function '_run_user_patches'`
- [ ] Migrate `_setup_apply_fixes_after_configuration()` from `setup-stack.sh` to `check-stack.sh` and change it to be registered as a check function instead of a setup function.
- [ ] Consider the same for `_setup_save_states()` (`startup/setup.d/mail_state.sh`), but take into account the that it's not destructive to a restart. If you prefer this could be split out to a separate task and tackled in future if lacking time to investigate.
- [ ] Adjust the conditional block to reflect the version proposed above which reduces the conditional down to calling `_setup`, as it looks like both `_early_supervisor_setup()` and `_check()` methods should be safe for restart.
- [ ] Optionally consider dropping `_early_supervisor_setup()` here and moving that method from `setup-stack.sh` to `daemons-stack.sh` where `_start_daemons()` can call it before the for loop.

polarathene reviewed (changes requested):
TLDR is a checklist of suggested changes at the end of one of the review comments.

Context for those change suggestions is covered by the rest of the feedback 👍

georglauterbach commented:
Alrighty - tests are currently running locally, but I think I finished the analysis and implementation for this one. Thank you for the feedback @casperklein and @polarathene - as always.

I have split everything into digestible commits, please review commit by commit.

@polarathene In your https://github.com/docker-mailserver/docker-mailserver/pull/4323#discussion_r1921304434, I have ticked what I implemented, and I explained what I had to do differently with comments in my commits.

polarathene reviewed (changes requested):
Summary of change requests from review feedback. Not everything needs to be addressed, but the env export needs to be resolved either as described below or by reverting that change (_it doesn't look like ENV can be adjusted for restarts, that results in a fresh container AFAIK, thus may not be a bug to restrict it to `_setup()`_)

- [x] `__environment_variables_export` is a regression now. To prevent that you'll need to:
  - [x] Migrate `_check_log_level()` from `check-stack.sh` into `variables-stack.sh`
  - [x] Migrate the feature flag co-located ldap, oauth, saslauthd `variable-stack.sh` method calls from `_register_functions` (`start-mailserver.sh`) to a new method in `variable-stack.sh` with equivalent feature flag ENV checks.
  - [ ] Have the two methods (_log-level check and feature-flag_) called prior to the export method.
    - **NOTE:** This would make the `_early_variables_setup` name a bit redundant, but I'm fine leaving that as-is for now in this PR so the diff is easier to grok / follow 😅 
- [ ] `_setup_adjust_state_permissions` and `_setup_directory_and_file_permissions` are both suitable to be managed during `_check`, but I won't enforce this change request for the PR.
  - **NOTE:** `_setup_adjust_state_permissions` is partially dependent upon `_setup_save_states` on a fresh container start should the volume not already provide the service dir for a feature. Running `_setup_save_states` in `_check` is also valid AFAIK, but the split is of `mail_state.sh` into two methods is still good to have.
- [ ] `_setup_adjust_state_permissions()` should inverse the debug log condition, as suggested. I don't see much value in debug log for notifying the user of skipping the logic, as per review feedback.

From prior feedback:
- [ ] Optionally consider dropping `_early_supervisor_setup()` from `start-mailserver.sh` and moving that method from `setup-stack.sh` to `daemons-stack.sh` where `_start_daemons()` can call it before the for loop.

georglauterbach commented on target/scripts/startup/setup.d/mail_state.sh:
```diff
@@ -3,123 +3,125 @@
+
+  local DEST DESTDIR SERVICEDIR SERVICEDIRS SERVICEFILE SERVICEFILES
+
+  # Always enabled features:
+  SERVICEDIRS=( 'lib/logrotate' 'lib/postfix' 'spool/postfix' )
```
I will revert this, came in with another diff I had laying around.

georglauterbach commented on target/scripts/startup/setup.d/mail_state.sh:
```diff
@@ -3,123 +3,125 @@
-    chmod 710 "${STATEDIR}/spool-postfix/public"
-  else
-    _log 'debug' "'${STATEDIR}' is not present; Not consolidating state"
+  if [[ ! -d ${DMS_STATE_DIR:?DMS_STATE_DIR is not set} ]]; then
+    _log 'debug' "'${DMS_STATE_DIR}' is not present - not consolidating state"
+    return 0
```
> This is to support sharing it across methods in `mail_state.sh` only right (_due to [`chore: split _setup_save_states for restart purposes`](https://github.com/docker-mailserver/docker-mailserver/pull/4323/commits/ece0f881b0f69410fa45bedf4f9c117927a5c5c8)_), because restarts are only running the ownership fixes portion (`_setup_adjust_state_permissions()`) of `mail_state.sh`?

Indeed, that is the case. If, at any point, we need this variable accidentally in another place, this will obviously also help. But, you are right, for now, it is only because of the split.

georglauterbach commented on target/scripts/startup/variables-stack.sh:
```diff
@@ -7,6 +7,18 @@ function _early_variables_setup() {
   _obtain_hostname_and_domainname
   __environment_variables_backwards_compatibility
   __environment_variables_general_setup
+  __environment_variables_export
```
Ah, you are right! I will move it to the end of the setup stack.

georglauterbach replied:
I opted to leave the `_early` prefix because I think it still very much fits the scheme: it needs to be called before everything else (except the supervisor setup, of course).

georglauterbach commented on target/scripts/start-mailserver.sh:
```diff
@@ -122,14 +122,17 @@ function _register_functions() {
   _register_setup_function '_setup_save_states'
-  _register_setup_function '_setup_apply_fixes_after_configuration'
-  _register_setup_function '_environment_variables_export'
+  _register_setup_function '_setup_adjust_state_permissions'
```
I'm not 100% sure what to make out of this now. Are you satisfied with the way it is or should I change something?

georglauterbach commented on target/scripts/start-mailserver.sh:
```diff
@@ -174,26 +177,24 @@ function _register_functions() {
+_check
 
-  _register_functions
+# Ensure DMS only adjusts config files for a new container.
+# Container restarts should skip as they retain the modified config.
+if [[ -f /CONTAINER_START ]]; then
+  _log 'info' 'Container was restarted. Skipping most setup routines.'
+  # We cannot skip all setup routines because some need to run _after_
+  # the initial setup (and hence, they cannot be moved to the check stack).
+  _setup_directory_and_file_permissions
+  _setup_adjust_state_permissions
+else
+  _setup
```
I have tried putting `_setup_directory_and_file_permissions` into `_check_`..., but this resulted in failing tests with `/var/mail/...` having wrong permissions. Due to a lack of time, I did not investigate further but assumed something during setup touched the directories. Hence, it needs to run after all other setup functions. Consequently, I was unable to move it into the check stack.

Please give it a try if you think I made a mistake elsewhere (entirely possible, really), but I think we cannot move it yet without more significant adjustments, which I'm reluctant towards considering we want to release a new major.

---

Concerning the log if `_setup_adjust_state_permissions`, I'll change that.

georglauterbach commented:
> Summary of change requests from review feedback. Not everything needs to be addressed, but the env export needs to be resolved either as described below or by reverting that change (_it doesn't look like ENV can be adjusted for restarts, that results in a fresh container AFAIK, thus may not be a bug to restrict it to `_setup()`_)
> 
> * [x]  `__environment_variables_export` is a regression now. To prevent that you'll need to:      
>   * [x]  Migrate `_check_log_level()` from `check-stack.sh` into `variables-stack.sh`
>   * [x]  Migrate the feature flag co-located ldap, oauth, saslauthd `variable-stack.sh` method calls from `_register_functions` (`start-mailserver.sh`) to a new method in `variable-stack.sh` with equivalent feature flag ENV checks.
>   * [x]  Have the two methods (_log-level check and feature-flag_) called prior to the export method.

Done

> * [ ]  `_setup_adjust_state_permissions` and `_setup_directory_and_file_permissions` are both suitable to be managed during `_check`, but I won't enforce this change request for the PR.
>   * **NOTE:** `_setup_adjust_state_permissions` is partially dependent upon `_setup_save_states` on a fresh container start should the volume not already provide the service dir for a feature. Running `_setup_save_states` in `_check` is also valid AFAIK, but the split is of `mail_state.sh` into two methods is still good to have.

I raised my concerns and problems with this in https://github.com/docker-mailserver/docker-mailserver/pull/4323#discussion_r1925969281.

>  * [ ]  `_setup_adjust_state_permissions()` should inverse the debug log condition, as suggested. I don't see much value in debug log for notifying the user of skipping the logic, as per review feedback.

Done (I think)
 
> From prior feedback:
> 
> * [ ]  Optionally consider dropping `_early_supervisor_setup()` from `start-mailserver.sh` and moving that method from `setup-stack.sh` to `daemons-stack.sh` where `_start_daemons()` can call it before the for loop.

I do not think this would be a viable approach, as the script calls `exit`, which I'd like to call as early as possible, and not after the setup stack has already completed.

georglauterbach commented on target/scripts/startup/setup.d/mail_state.sh:
```diff
@@ -3,123 +3,125 @@
+  if [[ ! -d ${DMS_STATE_DIR:?DMS_STATE_DIR is not set} ]]; then
+    _log 'debug' "'${DMS_STATE_DIR}' is not present - not adjusting state permissions"
+    return 0
+  fi
```
I have removed the log as per e549b73e090d1382a68673d9b30aac55d00c816c

polarathene commented on target/scripts/startup/setup.d/mail_state.sh:
```diff
@@ -3,123 +3,126 @@
+
+  for SERVICEFILE in "${SERVICEFILES[@]}"; do
+    DEST="${DMS_STATE_DIR}/${SERVICEFILE}"
+    DESTDIR="${DEST%/*}"
+
+    mkdir -p "${DESTDIR}"
```
Optional, but the variable seems to only be used for the `mkdir`, I find the comment a bit less ambiguous for those less familiar with what the bash syntax is doing:

```suggestion

    # Append service parent dir(s) path to the state dir and ensure it exists:
    mkdir -p "${DEST%/*}"
```

FWIW this does result in a path like `/var/mail-state//etc` (_with the double `/`_), non-issue for `mkdir -p` but might be an issue if used elsewhere? At least by dropping the variable name it's less prone to someone assuming the path is clean to use elsewhere? 🤔

polarathene commented on target/scripts/startup/setup.d/mail_state.sh:
```diff
@@ -3,123 +3,126 @@
+      _log 'trace' "Moving contents of ${SERVICEDIR} to ${DEST}"
+      # An empty volume was mounted, or new content dir now exists from enabling a feature ENV:
+      mv "${SERVICEDIR}" "${DEST}"
+      # Apply SELinux security context to match the state directory, so access
+      # is not restricted to the current running container:
+      chcon -R --reference="${DMS_STATE_DIR}" "${DEST}" 2>/dev/null || true
```
Inlining a reference since this PR will make `git blame` a lil annoying to find it otherwise 😅 (_I didn't suggest this for the earlier duplicate comment, since that's just a single file and for the most part it looks like this logic could be DRY as a function call if someone wants to refactor_).

```suggestion
      # Apply SELinux security context to match the state directory, so access
      # is not restricted to the current running container:
      # https://github.com/docker-mailserver/docker-mailserver/pull/3890
      chcon -R --reference="${DMS_STATE_DIR}" "${DEST}" 2>/dev/null || true
```

polarathene reviewed (changes requested):
Apologies for the long delay getting back to this. Looking good, just some minor changes.

---

> I do not think this would be a viable approach, as [the script calls `exit`](https://github.com/docker-mailserver/docker-mailserver/blob/4d8a56072a88cbacea373be88aa1d805e1eacc59/target/scripts/startup/setup-stack.sh#L47-L48), which I'd like to call as early as possible, and not after the setup stack has already completed.

That makes sense, thanks for pointing that out! :)

---

> > * [ ]   `_setup_adjust_state_permissions` and `_setup_directory_and_file_permissions` are both suitable to be managed during `_check`, but I won't enforce this change request for the PR.
> >   
> >   * **NOTE:** `_setup_adjust_state_permissions` is partially dependent upon `_setup_save_states` on a fresh container start should the volume not already provide the service dir for a feature. Running `_setup_save_states` in `_check` is also valid AFAIK, but the split is of `mail_state.sh` into two methods is still good to have.
> 
> I raised my concerns and problems with this in [#4323 (comment)](https://github.com/docker-mailserver/docker-mailserver/pull/4323#discussion_r1925969281).

I'm not quite sure why the `mail-state.sh` methods for `/var/mail-state` would cause some breakage that affects `/var/mail` from that proposed change. I'll look into trying the change myself to reproduce 🤔 

I won't block on that though, we'll get this PR out of the way first :)

georglauterbach commented:
I applied the changes and appeased the linting gods.

casperklein commented:
Some update from my side: This went from a small change into something bigger + a lot of discussion which I haven't fully read yet. To make a proper review, I'll have to take some time. At the moment, that's a bit challenging for me. I'd like to review asap, but unfortunately I can't give an ETA right now.
If this can wait some more, I am happy to review. Otherwise I try to review afterwards. I'll leave it to you both, when this gets merged 👍

polarathene commented:
> If this can wait some more, I am happy to review. Otherwise I try to review afterwards. I'll leave it to you both, when this gets merged

I'm in no rush, I'll be working on the podman PR in the meantime 👍 

> This went from a small change into something bigger + a lot of discussion which I haven't fully read yet.

If it helps here is a quick overview for the diff.

- **Primary change** is in `start-mailserver.sh` to adjust the restart logic, it's mostly unified now so that it only calls `_setup` on fresh containers.
- **Additional changes** are more house-keeping related, some supportive to the primary change. A bulk of the diffs aren't as large as they seem, just the shuffling of lines and indentation adjustments.

---

**Primary change** is adjusting the restart check for `/CONTAINER_START` to this:

```bash
_early_supervisor_setup
_early_variables_setup

_log 'info' "Welcome to docker-mailserver ${DMS_RELEASE}"

_register_functions
_check

# Ensure DMS only adjusts config files for a new container.
# Container restarts should skip as they retain the modified config.
if [[ -f /CONTAINER_START ]]; then
  _log 'info' 'Container was restarted. Skipping most setup routines.'
  # We cannot skip all setup routines because some need to run _after_
  # the initial setup (and hence, they cannot be moved to the check stack).
  _setup_directory_and_file_permissions
  _setup_adjust_state_permissions
else
  _setup
fi
```

- Effectively now a restart runs everything except `_setup`.
- There are two `_setup` methods that still need to be run for restarts, but were proposed to be migrated to `_check` functions.
  - `_setup_directory_and_file_permissions()` is a bunch of misc cleanup operations. One of those is to remove PID files (_primary regression we're fixing by ensuring this method is run_), and also ensures `/var/mail` ownership is corrected if necessary.
  - `_setup_adjust_state_permissions()` is a new method in `mail_state.sh`, which was split into two separate methods. This is the 2nd method for `mail_state.sh` which just manages ownership/permission fixes, while the earlier portion only runs on fresh containers to create/copy and symlink to the DMS state volume (_if any ENV is changed to toggle a feature, Docker will force recreate the container, however it looks like it's probably safe to still run this during restarts too, we effectively were prior to v14_).
- @georglauterbach did attempt to migrate those methods to `_check()`, but ran into test failures with `/var/mail` ownership. I haven't looked into reproducing that yet. That would simplify the conditional block, removing the two separate calls for restart there and also their repeated use in `_setup`. I'm not too fond of the two call sites and would prefer to simplify, but not blocking the PR on that.

**Additional changes:**
- `STATEDIR` in `mail_state.sh` was renamed to `DMS_STATE_DIR`, along with an adjust to the conditional and related log and that is repeated with the 2nd method.
  - Part of this change removed the indent of the method as it previously was all wrapped in the conditional (_that's now inverted to exit early_).
  - It also introduced `__declare_readonly()`, which personally I don't feel is necessary and adds some complexity, but I can understand why @georglauterbach wanted to introduce it (_this is more of a DX issue with using bash though_). It could be avoided and `DMS_STATE_DIR` could just be declared at the top of `mail_state.sh`.
- While the diff is large/noisy for `mail_state.sh` actual changes is minimal and it's effectively white-space. I did add some minor housekeeping changes but dropping a variable name `DESTDIR` that was only used for `mkdir -p`, preferring to inline that and add a comment instead. Likewise for a `chcon` call, I had the comment append a reference link in favor over relying on `git blame` for context.
- As part of the earlier restart logic changes, `_setup()` was making calls to setup ENV vars via `variables-stack.sh` but most of these were not registered as part of `_setup()`, just co-located with the equivalent feature. These have instead been extracted out and managed via `variables-stack.sh` at an earlier stage as part of `_early_variables_setup()`.
- `_check_log_level()` (`check-stack.sh`) was migrated to `variables-stack.sh` as `__environment_variables_log_level()`, which is a house-keeping thing rather than addressing anything in particular, the logic was ENV related so it belonged in `variables-stack.sh` instead.

georglauterbach commented:
Indeed, this has become bigger than I anticipated as well. I'd still like to merge ASAP TBH, because I'd reall like publish v15.0.0.

Here is my proposal: I'd merge this and prepare the v15 release. We'll pick 1.5 weeks for the freeze because of this change. Please let me know whether this is okay for you both.

---

@polarathene thanks for the overview :) I added the `__declare_readonly` to ensure we do not run into problems with `readonly` variables being defined more than once. It's really a Bash thingy, I guess..

georglauterbach commented:
I have resolved the merge conflicts with `master` and enabled autp-merge now.

polarathene commented on CHANGELOG.md:
```diff
@@ -19,8 +19,8 @@ All notable changes to this project will be documented in this file. The format
     - To preserve this state when upgrading to DMS v15, **you must manually migrate `getmail/` from the _DMS Config Volume_ to `lib-getmail/` in the _DMS State Volume_.**
-  - Added missing `debug getmail` subcommand to `setup` ([#4346](https://github.com/docker-mailserver/docker-mailserver/pull/4346))
-- **removed `VERSION`** file that was used for checking version updates ([#3677](https://github.com/docker-mailserver/docker-mailserver/issues/3677),[#4321](https://github.com/docker-mailserver/docker-mailserver/pull/4321))
+- **removed `VERSION` file** that was used for checking version updates ([#3677](https://github.com/docker-mailserver/docker-mailserver/issues/3677),[#4321](https://github.com/docker-mailserver/docker-mailserver/pull/4321))
+- **updated container restart behavior** to include checks and more permission adjustments ([#4323](https://github.com/docker-mailserver/docker-mailserver/pull/4323))
```
```suggestion
- **Removed `VERSION` file** that was used for checking version updates ([#3677](https://github.com/docker-mailserver/docker-mailserver/issues/3677),[#4321](https://github.com/docker-mailserver/docker-mailserver/pull/4321))
- **Updated container restart behavior** to include checks and more permission adjustments ([#4323](https://github.com/docker-mailserver/docker-mailserver/pull/4323))
```

I'm not sure that either of these are considered breaking?

- `VERSION` removal is breaking in the sense of it affects all releases [prior to v13](https://github.com/docker-mailserver/docker-mailserver/pull/3666).
- The updated restart behaviour is more of a regression fix from v14 changes.

polarathene commented on CHANGELOG.md:
```diff
@@ -31,6 +31,7 @@ All notable changes to this project will be documented in this file. The format
 
 - **Internal:**
   - Add password confirmation to several `setup` CLI subcommands ([#4072](https://github.com/docker-mailserver/docker-mailserver/pull/4072))
+- `debug getmail` subcommand to `setup` ([#4346](https://github.com/docker-mailserver/docker-mailserver/pull/4346))
```
```suggestion
  - Added a `debug getmail` subcommand to `setup` ([#4346](https://github.com/docker-mailserver/docker-mailserver/pull/4346))
```

polarathene reviewed (changes requested):
I'll apply these and shift the changelog breaking items to more appropriate sections.

casperklein commented:
@georglauterbach Good structured commits 👍 
@polarathene Thank you for the summary, that helped me to understand things (faster). 👍 

The only nitpick I have is the "DMS_STATE_DIR" change:

- Why making it global when not needed?
- DMS_STATE_DIR is set explicitly, why is this check needed? `{DMS_STATE_DIR:?DMS_STATE_DIR is not set}`. This should never be unset?
- do we really need `__declare_readonly`?

I just think, that this makes things unnecessary complicated. Like @polarathene said, `DMS_STATE_DIR could just be declared at the top of mail_state.sh` should be enough.

polarathene commented:
> Why making it global when not needed?

I felt the same way but I didn't want to nitpick heavily on it.

I understand the intention to minimize caveats from using Bash. It's one of the main friction points we have with maintaining the project when variables / methods are being pulled into scope from other scripts, something we shouldn't have to feel like we're walking on glass with.

If we do explore migrating some scripts away from bash at some point, that's one of the main benefits. So far NuShell seems like a good compromise for that (_a rust-based shell that provides optional type annotation, structured data, module imports, etc_). I understand wanting to avoid Python and concerns with languages that'd involve compilation vs an interpreter for scripts, Nu might be a good candidate for such. We've had this discussion quite a bit so I won't drag it out again here 😅 

I do agree that I'm not fond of the added complexity / abstraction to add preventative measures prematurely though.

georglauterbach commented:
I will remove this part again after seeing your feedback :)
